# Preprocessing

In [38]:
# Import dependencies
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import AgglomerativeClustering
from sklearn.cluster import KMeans
import hvplot.pandas
import plotly.express as px
import plotly.figure_factory as ff

In [39]:
# Load the data
file_path = "./Resources/crypto_data.csv"
crypto_df = pd.read_csv(file_path)
crypto_df.head(10)

Unnamed: 0   CoinName Algorithm  IsTrading ProofType  TotalCoinsMined  \
0         42    42 Coin    Scrypt       True   PoW/PoS     4.199995e+01   
1        365    365Coin       X11       True   PoW/PoS              NaN   
2        404    404Coin    Scrypt       True   PoW/PoS     1.055185e+09   
3        611  SixEleven   SHA-256       True       PoW              NaN   
4        808        808   SHA-256       True   PoW/PoS     0.000000e+00   
5       1337  EliteCoin       X13       True   PoW/PoS     2.927942e+10   
6       2015  2015 coin       X11       True   PoW/PoS              NaN   
7        BTC    Bitcoin   SHA-256       True       PoW     1.792718e+07   
8        ETH   Ethereum    Ethash       True       PoW     1.076842e+08   
9        LTC   Litecoin    Scrypt       True       PoW     6.303924e+07   

  TotalCoinSupply  
0              42  
1      2300000000  
2       532000000  
3          611000  
4               0  
5    314159265359  
6               0  
7        21000000  
8               0  
9        84000000

In [40]:
crypto_df.count()
crypto_df.shape

(1252, 7)

In [41]:
crypto_df.dtypes

Unnamed: 0          object
CoinName            object
Algorithm           object
IsTrading             bool
ProofType           object
TotalCoinsMined    float64
TotalCoinSupply     object
dtype: object

In [42]:
crypto_df.columns

Index(['Unnamed: 0', 'CoinName', 'Algorithm', 'IsTrading', 'ProofType',
       'TotalCoinsMined', 'TotalCoinSupply'],
      dtype='object')

In [43]:
# STEP 1
# Remove all cryptocurrencies that aren’t trading
#crypto_df[crypto_df["IsTrading"] == True].count()
crypto_df.drop(crypto_df[crypto_df["IsTrading"] == False].index, inplace=True)
crypto_df.shape

(1144, 7)

108 cryptocurrencies were removed because they are not trading, leaving the new dataframe with 1144 cryptocurrencies

In [44]:
# STEP 2
# Remove all cryptocurrencies that don’t have an algorithm defined
for column in crypto_df.columns:
    print(f"Column {column} has {crypto_df[column].isnull().sum()} null values")

Column Unnamed: 0 has 0 null values
Column CoinName has 0 null values
Column Algorithm has 0 null values
Column IsTrading has 0 null values
Column ProofType has 0 null values
Column TotalCoinsMined has 459 null values
Column TotalCoinSupply has 0 null values


When we explore the dataset for null values we see that all cryptocurrencies have a defined algorithm, meaning there are no null values for the Algorithm column. 

In [45]:
# STEP 3
# Remove the IsTrading column.
crypto_df.drop(columns=["IsTrading"], inplace=True)
crypto_df.head()

Unnamed: 0   CoinName Algorithm ProofType  TotalCoinsMined TotalCoinSupply
0         42    42 Coin    Scrypt   PoW/PoS     4.199995e+01              42
1        365    365Coin       X11   PoW/PoS              NaN      2300000000
2        404    404Coin    Scrypt   PoW/PoS     1.055185e+09       532000000
3        611  SixEleven   SHA-256       PoW              NaN          611000
4        808        808   SHA-256   PoW/PoS     0.000000e+00               0

In [46]:
# STEP 4
# Remove all cryptocurrencies with at least one null value.
crypto_df = crypto_df.dropna()
crypto_df.shape

(685, 6)

We can see from an exploratory analysis we did during Step 2 that there were 459 null values in the TotalCoinsMined column. Those 459 rows were dropped, leaving us with 685 rows.

In [47]:
# STEP 5
# Remove all cryptocurrencies without coins mined.
crypto_df.drop(crypto_df[crypto_df["TotalCoinsMined"] <= 0].index, inplace=True)

In [48]:
crypto_df.head()

Unnamed: 0   CoinName Algorithm ProofType  TotalCoinsMined TotalCoinSupply
0         42    42 Coin    Scrypt   PoW/PoS     4.199995e+01              42
2        404    404Coin    Scrypt   PoW/PoS     1.055185e+09       532000000
5       1337  EliteCoin       X13   PoW/PoS     2.927942e+10    314159265359
7        BTC    Bitcoin   SHA-256       PoW     1.792718e+07        21000000
8        ETH   Ethereum    Ethash       PoW     1.076842e+08               0

Finding and removies all cryptocurrencies with less than or equal to 0 coins mined; leaving our dataframe with 532 cryptocurrencies.

In [49]:
# Create a cleaned dataframe
clean_crypto_df = crypto_df
clean_crypto_df.head()

Unnamed: 0   CoinName Algorithm ProofType  TotalCoinsMined TotalCoinSupply
0         42    42 Coin    Scrypt   PoW/PoS     4.199995e+01              42
2        404    404Coin    Scrypt   PoW/PoS     1.055185e+09       532000000
5       1337  EliteCoin       X13   PoW/PoS     2.927942e+10    314159265359
7        BTC    Bitcoin   SHA-256       PoW     1.792718e+07        21000000
8        ETH   Ethereum    Ethash       PoW     1.076842e+08               0

In [50]:
clean_crypto_df.shape

(532, 6)

In [51]:
# STEP 6
# Store the names of all cryptocurrencies on a DataFramed named coins_name, 
# and use the crypto_df.index as the index for this new DataFrame.
coins_name = pd.DataFrame(clean_crypto_df[["Unnamed: 0", "CoinName"]])
coins_name.set_index("Unnamed: 0", drop=True, inplace=True)
coins_name

CoinName
Unnamed: 0                  
42                   42 Coin
404                  404Coin
1337               EliteCoin
BTC                  Bitcoin
ETH                 Ethereum
LTC                 Litecoin
DASH                    Dash
XMR                   Monero
ETC         Ethereum Classic
ZEC                    ZCash
BTS                Bitshares
DGB                 DigiByte
BTCD             BitcoinDark
XPY                  PayCoin
PRC              ProsperCoin
KOBO                KoboCoin
SPR               Spreadcoin
ARG                 Argentum
AUR              Aurora Coin
BLU                 BlueCoin
XMY               MyriadCoin
MOON                MoonCoin
ZET                 ZetaCoin
SXC                  SexCoin
QTL                  Quatloo
ENRG              EnergyCoin
QRK                QuarkCoin
RIC                  Riecoin
DGC             Digitalcoin 
BTB                   BitBar
...                      ...
SMLY              SmileyCoin
MI                XiaoMiCoin
GRS              Groestlcoin
CPC                CapriCoin
CLUB                ClubCoin
RADS                  Radium
BTA                     Bata
PAK                  Pakcoin
CRB               Creditbit 
OK                    OKCash
LSK                     Lisk
XHI                   HiCoin
XWC                WhiteCoin
FSC           FriendshipCoin
J                   JoinCoin
TRI           Triangles Coin
VOLLAR                Vollar
EOS                      EOS
RCC            Reality Clash
OXY                  Oxycoin
TCH                TigerCash
WAVES                  Waves
PART                 Particl
BTT               BitTorrent
NXT                      Nxt
ZEPH                  ZEPHYR
GAP                  Gapcoin
BDX                   Beldex
ZEN                  Horizen
XBC              BitcoinPlus

[532 rows x 1 columns]

In [52]:
# STEP 7
# Remove the CoinName column from the clean_crypto_df
clean_crypto_df.drop(columns=["CoinName"], inplace=True)
clean_crypto_df

Unnamed: 0        Algorithm       ProofType  TotalCoinsMined  \
0            42           Scrypt         PoW/PoS     4.199995e+01   
2           404           Scrypt         PoW/PoS     1.055185e+09   
5          1337              X13         PoW/PoS     2.927942e+10   
7           BTC          SHA-256             PoW     1.792718e+07   
8           ETH           Ethash             PoW     1.076842e+08   
9           LTC           Scrypt             PoW     6.303924e+07   
10         DASH              X11         PoW/PoS     9.031294e+06   
11          XMR   CryptoNight-V7             PoW     1.720114e+07   
12          ETC           Ethash             PoW     1.133597e+08   
13          ZEC         Equihash             PoW     7.383056e+06   
14          BTS          SHA-512             PoS     2.741570e+09   
15          DGB         Multiple             PoW     1.140622e+10   
16         BTCD          SHA-256         PoW/PoS     1.288862e+06   
19          XPY          SHA-256             PoS     1.199533e+07   
20          PRC           Scrypt             PoW     5.768310e+06   
24         KOBO              X15         PoW/PoS     2.554254e+07   
27          SPR              X11             PoW     1.114973e+07   
37          ARG           Scrypt             PoW     1.220580e+07   
39          AUR           Scrypt         PoW/PoS     1.813510e+07   
42          BLU           Scrypt         PoW/PoS     6.354239e+08   
44          XMY         Multiple             PoW     1.688254e+09   
45         MOON           Scrypt             PoW     8.800000e+01   
46          ZET          SHA-256             PoW     1.702044e+08   
47          SXC           Scrypt             PoW     1.289410e+08   
48          QTL           Scrypt             PoW     7.357795e+06   
49         ENRG           Scrypt         PoW/PoS     1.230628e+08   
50          QRK            Quark             PoW     2.595572e+08   
51          RIC          Groestl             PoW     4.954859e+07   
52          DGC           Scrypt             PoW     3.344299e+07   
54          BTB           Scrypt         PoW/PoS     4.257948e+04   
...         ...              ...             ...              ...   
1181       SMLY           Scrypt             PoW     2.958399e+10   
1185         MI           Scrypt         PoW/PoS     4.043165e+08   
1186        GRS          Groestl             PoW     7.329820e+07   
1187        CPC              X11         PoW/PoS     2.013643e+08   
1188       CLUB           Scrypt         PoW/PoS     1.036219e+08   
1189       RADS              PoS             PoS     3.821246e+06   
1190        BTA           Scrypt             PoW     5.052556e+06   
1191        PAK           Scrypt             PoW     6.968446e+07   
1192        CRB              X11  Proof of Trust     1.690102e+07   
1193         OK          SHA-256         PoW/PoS     7.467606e+07   
1194        LSK             DPoS            DPoS     1.200121e+08   
1195        XHI           Scrypt             PoS     1.000024e+10   
1196        XWC           Scrypt         PoW/PoS     2.520056e+08   
1198        FSC        NeoScrypt         PoW/PoS     1.120385e+06   
1201          J         Multiple             PoW     3.332922e+06   
1202        TRI              X13         PoW/PoS     1.407778e+05   
1204     VOLLAR  Equihash+Scrypt             PoW     1.000000e+08   
1211        EOS             DPoS            DPoS     1.020545e+09   
1213        RCC           Ethash             PoW     2.448794e+07   
1214        OXY             DPoS            DPoS     1.122382e+09   
1226        TCH          SHA-256             PoS     1.000000e+09   
1230      WAVES       Leased POS            LPoS     1.000000e+08   
1231       PART              PoS             PoS     9.283138e+06   
1234        BTT            TRC10            DPoS     9.899887e+11   
1237        NXT              PoS        PoS/LPoS     1.000000e+09   
1238       ZEPH          SHA-256            DPoS     2.000000e+09   
124

In [53]:
# STEP 8
# Create dummies variables for all of the text features, and store the resulting data on a DataFrame named X.

# Look at datatypes in coins_name
clean_crypto_df.dtypes

Unnamed: 0          object
Algorithm           object
ProofType           object
TotalCoinsMined    float64
TotalCoinSupply     object
dtype: object

In [54]:
# First, change TotalCoinSupply datatype from object to float
clean_crypto_df["TotalCoinSupply"] = clean_crypto_df["TotalCoinSupply"].apply(float)
clean_crypto_df.dtypes

Unnamed: 0          object
Algorithm           object
ProofType           object
TotalCoinsMined    float64
TotalCoinSupply    float64
dtype: object

In [55]:
# Next, we want to convert the ProofType column to numbers
# First, let's see what ProofTypes are the most common
# proofType_df = clean_crypto_df.groupby("ProofType").nunique()
# print(proofType_df)

POS, POW, and POW/POS contain the most types of cryptocurrencies. So we can group those 3 individually and the rest as 'other'.

In [56]:
# Transform the string column.
# def change_string(ProofType):
#     if ProofType == "PoS":
#         return 0
#     if ProofType == "PoW":
#         return 1
#     if ProofType == "PoW/PoS":
#         return 2
#     else:
#         return 3
    
# coins_name["ProofType"] = coins_name["ProofType"].apply(change_string)
# coins_name

In [57]:
# Next create dummy variables for all text features (Algorithm and ProofType)
X = clean_crypto_df[["Algorithm", "ProofType", "TotalCoinsMined", "TotalCoinSupply"]].copy()
X = pd.get_dummies(X, columns=["Algorithm", "ProofType"], drop_first=True)
X = X.dropna()
X.head()

TotalCoinsMined  TotalCoinSupply  Algorithm_536  Algorithm_Argon2d  \
0     4.199995e+01     4.200000e+01              0                  0   
2     1.055185e+09     5.320000e+08              0                  0   
5     2.927942e+10     3.141593e+11              0                  0   
7     1.792718e+07     2.100000e+07              0                  0   
8     1.076842e+08     0.000000e+00              0                  0   

   Algorithm_BLAKE256  Algorithm_Blake  Algorithm_Blake2S  Algorithm_Blake2b  \
0                   0                0                  0                  0   
2                   0                0                  0                  0   
5                   0                0                  0                  0   
7                   0                0                  0                  0   
8                   0                0                  0                  0   

   Algorithm_C11  Algorithm_Cloverhash         ...          ProofType_PoW/PoS  \
0              0                     0         ...                          1   
2              0                     0         ...                          1   
5              0                     0         ...                          1   
7              0                     0         ...                          0   
8              0                     0         ...                          0   

   ProofType_PoW/PoS   ProofType_PoW/PoW  ProofType_PoW/nPoS  ProofType_Pos  \
0                   0                  0                   0              0   
2                   0                  0                   0              0   
5                   0                  0                   0              0   
7                   0                  0                   0              0   
8                   0                  0                   0              0   

   ProofType_Proof of Authority  ProofType_Proof of Trust  ProofType_TPoS  \
0                             0                         0               0   
2                             0                         0               0   
5                             0                         0               0   
7                             0                         0               0   
8                             0                         0               0   

   ProofType_Zero-Knowledge Proof  ProofType_dPoW/PoW  
0                               0                   0  
2                               0                   0  
5                               0                   0  
7                               0                   0  
8                               0                   0  

[5 rows x 96 columns]

In [58]:
# STEP 9
# Use the StandardScaler from sklearn to standardize all of the data from the X DataFrame.
X_scaled = StandardScaler().fit_transform(X)
print(X_scaled[0:2])

[[-0.11710817 -0.1528703  -0.0433963  -0.0433963  -0.06142951 -0.07530656
  -0.0433963  -0.06142951 -0.06142951 -0.0433963  -0.0433963  -0.19245009
  -0.06142951 -0.09740465 -0.0433963  -0.11547005 -0.07530656 -0.0433963
  -0.0433963  -0.15191091 -0.0433963  -0.13118084 -0.0433963  -0.0433963
  -0.08703883 -0.0433963  -0.0433963  -0.0433963  -0.0433963  -0.06142951
  -0.0433963  -0.08703883 -0.08703883 -0.08703883 -0.0433963  -0.13118084
  -0.13840913 -0.13840913 -0.0433963  -0.06142951 -0.0433963  -0.07530656
  -0.18168574 -0.0433963  -0.0433963  -0.0433963  -0.07530656 -0.15826614
  -0.31491833 -0.0433963  -0.08703883 -0.07530656 -0.06142951  1.38675049
  -0.0433963  -0.0433963  -0.06142951 -0.0433963  -0.0433963  -0.0433963
  -0.0433963  -0.0433963  -0.0433963  -0.0433963  -0.0433963  -0.39879994
  -0.0433963  -0.18168574 -0.0433963  -0.08703883 -0.08703883 -0.10680283
  -0.13118084 -0.0433963  -0.0433963  -0.0433963  -0.0433963  -0.07530656
  -0.43911856 -0.0433963  -0.06142951 -0.

In [59]:
# Save cleaned data
file_path = "./Resources/X_cleaned_crypto.csv"
X.to_csv(file_path, index=False)

# Reducing Data Dimensions Using PCA

In [60]:
# Initialize PCA model
pca = PCA(n_components=3)

In [61]:
# Get three principal components for the X_scaled data.
X_pca = pca.fit_transform(X_scaled)

In [62]:
# Transform PCA data to a dataframe
X_pca_df = pd.DataFrame(
    data=X_pca, columns=["PC 1", "PC 2", "PC 3"]
)
X_pca_df.head()

PC 1      PC 2      PC 3
0 -0.333955  0.993608 -0.522746
1 -0.317289  0.993657 -0.523065
2  2.294921  1.563401 -0.619588
3 -0.147424 -1.275470  0.171421
4 -0.151750 -2.044473  0.356146

In [63]:
# Fetch the explained variance
pca.explained_variance_ratio_

array([0.02851297, 0.02177838, 0.02092444])

In [64]:
# Join the index from crypto_df
join_df = pd.DataFrame(data=clean_crypto_df, columns=["Unnamed: 0"])
pca_df = X_pca_df.join(join_df, how="inner")
pca_df.head()

PC 1      PC 2      PC 3 Unnamed: 0
0 -0.333955  0.993608 -0.522746         42
2  2.294921  1.563401 -0.619588        404
5 -0.162368 -1.160095  0.015884       1337
7 -0.148343 -2.231451  0.415101        BTC
8 -0.150192 -2.044587  0.356132        ETH

In [65]:
# Remove "Unnamed: 0"
pca_df = pca_df.set_index(["Unnamed: 0"])
pca_df

PC 1      PC 2       PC 3
Unnamed: 0                                
42          -0.333955  0.993608  -0.522746
404          2.294921  1.563401  -0.619588
1337        -0.162368 -1.160095   0.015884
BTC         -0.148343 -2.231451   0.415101
ETH         -0.150192 -2.044587   0.356132
LTC         -0.154867 -2.041512   0.432079
DASH        -0.282779  1.347116  -0.020772
XMR          0.167585 -1.748403   0.321259
ETC         -0.317841  0.878202  -0.367228
ZEC         -0.271402  0.607574  -0.062275
BTS         -0.163515 -1.160079   0.015904
DGB         -0.251787  1.909616  -0.606535
BTCD        -0.226992 -0.788140   0.032171
XPY         -0.326189  0.993815  -0.522920
PRC         -0.085857 -1.741028   0.325015
KOBO        -0.162936 -1.160122   0.015899
SPR         -0.204840 -1.199045   0.111780
ARG          0.230495 -1.280998   0.165082
AUR         -0.163330 -1.160074   0.015900
BLU         -0.351652  1.400049  -0.090706
XMY         -0.163641 -1.160075   0.015906
MOON        -0.179137 -1.644759   0.228306
ZET         -0.142860 -1.160569   0.015583
SXC         -0.162436 -1.160087   0.015885
QTL         -0.162713 -1.160096   0.015891
ENRG        -0.332013  0.993571  -0.522778
QRK         -0.330602  1.095455  -0.201980
RIC          1.585623 -1.180819  -0.014087
DGC         -0.397119  1.365539  -0.506484
BTB         -0.160002 -1.160167   0.015850
...               ...       ...        ...
FRWC        -0.227060 -0.788143   0.032173
MUDRA       -0.365951  0.934344  -0.351235
PIZZA       -0.359023  1.398068  -0.184589
LC          -0.163560 -1.160081   0.015905
KMD         -0.135958 -2.232422   0.414999
GB          -0.157625 -2.340034   0.429153
CTC         -0.284419  0.722863  -0.217835
KRB         -0.193409  0.983969  -0.524081
BXT         -0.200782 -1.200417   0.190011
ZYD         -0.146906  0.242026   0.154244
MST         -0.049477 -2.044339   0.430484
ZNE         -0.152580 -2.268084   0.365433
COVAL       -0.147424 -1.275470   0.171420
ROOT         0.379906 -2.367758   0.384120
DOPE        -0.406101  1.668720  -0.489534
BTCR        -0.154579 -2.041512   0.432073
DLC         -0.181510 -1.484129   0.149906
SCN         -0.350966  1.094898  -0.201523
SH          -0.295906  1.883833  -0.496210
BUZZ        -0.305069  1.893032  -0.732511
XPO         -0.119559 -0.425541  -0.054699
PSB         -0.145567 -1.275529   0.171394
HALLO       -0.101154 -2.235824   0.409761
XZC         -0.204401 -1.199049   0.111773
CFC         -0.161593 -1.160148   0.015877
ROYAL       -0.120605 -1.740469   0.325591
GNJ         -0.252328  0.607349  -0.062602
TEAM        -0.417614  3.430502  13.911529
LANA        34.066598  1.579395  -0.756369
ELE         -0.346040  1.366557  -0.115255

[205 rows x 3 columns]

In [66]:
# Save the PC data
file_path = "./Resources/principal_components.csv"
pca_df.to_csv(file_path, index=False)

# Clustering Cryptocurrencies Using K-Means

In [67]:
# STEP 1
# Create an elbow curve to find the best value for K, and use the pcs_df DataFrame.

# Create empty lists
inertia = []
k = list(range(1, 11))

# Calculate the inertia for the range of K values
for i in k:
   km = KMeans(n_clusters=i, random_state=0)
   km.fit(pca_df)
   inertia.append(km.inertia_)

# Create the Elbow Curve using hvPlot
elbow_data = {"k": k, "inertia": inertia}
df_elbow = pd.DataFrame(elbow_data)
df_elbow.hvplot.line(x="k", y="inertia", xticks=k, title="Elbow Curve")

:Curve   [k]   (inertia)

Based off the elbow curve, I'd say that our ideal number of clusters is 4. 

In [68]:
km = KMeans(n_clusters=4)
km.fit(pca_df)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=4, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=None, tol=0.0001, verbose=0)

In [69]:
# STEP 2
# Once you define the best value for K, run the K-means algorithm to predict the K clusters for the cryptocurrencies’ data.
# Use the pcs_df to run the K-means algorithm.

# Initialize the K-means model
model = KMeans(n_clusters=4, random_state=0)

# Fit the model
model.fit(pca_df)

# Predict clusters
predictions = model.predict(pca_df)
print(predictions)

[0 0 2 2 2 2 0 2 0 0 2 0 2 0 2 2 2 2 2 0 2 2 2 2 2 0 0 2 0 2 2 0 0 2 2 0 0
 2 0 0 2 2 2 0 2 2 2 0 0 2 0 0 0 2 0 2 2 2 2 0 0 2 2 2 0 0 2 0 0 2 2 0 2 2
 0 0 0 0 0 2 0 0 2 0 0 2 0 2 0 2 0 0 0 0 0 0 2 2 0 0 0 2 0 0 2 0 0 0 0 0 2
 2 0 0 0 0 2 0 2 2 0 0 0 0 2 0 0 0 2 2 0 0 2 2 0 2 0 2 0 0 2 2 0 0 0 2 2 2
 0 0 0 0 2 0 0 2 0 2 0 0 0 2 0 2 2 0 2 0 2 0 0 2 2 0 0 2 0 0 2 2 2 0 0 2 0
 2 2 2 2 0 2 2 0 0 0 2 2 2 2 2 2 0 3 1 0]


In [70]:
# STEP 3
# Create a new DataFrame named “clustered_df,” that includes the following columns:
# Algorithm, ProofType, TotalCoinsMined, TotalCoinSupply, PC 1, PC 2, PC 3, CoinName, and Class. 

clustered_df = clean_crypto_df.merge(pca_df, on="Unnamed: 0")
clustered_df = clustered_df.merge(coins_name, on="Unnamed: 0")

clustered_df["class"] = model.labels_

clustered_df.set_index("Unnamed: 0", drop=True, inplace=True)
clustered_df.head(10)

Algorithm ProofType  TotalCoinsMined  TotalCoinSupply  \
Unnamed: 0                                                               
42                  Scrypt   PoW/PoS     4.199995e+01     4.200000e+01   
404                 Scrypt   PoW/PoS     1.055185e+09     5.320000e+08   
1337                   X13   PoW/PoS     2.927942e+10     3.141593e+11   
BTC                SHA-256       PoW     1.792718e+07     2.100000e+07   
ETH                 Ethash       PoW     1.076842e+08     0.000000e+00   
LTC                 Scrypt       PoW     6.303924e+07     8.400000e+07   
DASH                   X11   PoW/PoS     9.031294e+06     2.200000e+07   
XMR         CryptoNight-V7       PoW     1.720114e+07     0.000000e+00   
ETC                 Ethash       PoW     1.133597e+08     2.100000e+08   
ZEC               Equihash       PoW     7.383056e+06     2.100000e+07   

                PC 1      PC 2      PC 3          CoinName  class  
Unnamed: 0                                                         
42         -0.333955  0.993608 -0.522746           42 Coin      0  
404         2.294921  1.563401 -0.619588           404Coin      0  
1337       -0.162368 -1.160095  0.015884         EliteCoin      2  
BTC        -0.148343 -2.231451  0.415101           Bitcoin      2  
ETH        -0.150192 -2.044587  0.356132          Ethereum      2  
LTC        -0.154867 -2.041512  0.432079          Litecoin      2  
DASH       -0.282779  1.347116 -0.020772              Dash      0  
XMR         0.167585 -1.748403  0.321259            Monero      2  
ETC        -0.317841  0.878202 -0.367228  Ethereum Classic      0  
ZEC        -0.271402  0.607574 -0.062275             ZCash      0

# Visualizing Results

In [71]:
# STEP 1 
# Create a 3D scatter plot using Plotly Express to plot the clusters using the clustered_df DataFrame.

# Plot the df with a third axis to show what's happening with points in the middle
fig = px.scatter_3d(
    clustered_df,
    x="PC 1",
    y="PC 2",
    z="PC 3",
    color="class",
    symbol="class",
    hover_name="CoinName",
    hover_data=["Algorithm"],
    width=800,
)

fig.update_layout(legend=dict(x=0, y=1))
fig.show()

In [72]:
# STEP 2 
# Use hvplot.table to create a data table with all the current tradable cryptocurrencies.

clustered_df.hvplot.table(columns=["CoinName", "Algorithm", "ProofType", "TotalCoinSupply", "TotalCoinsMined", "class"], width=800)

:Table   [CoinName,Algorithm,ProofType,TotalCoinSupply,TotalCoinsMined,class]

In [73]:
# STEP 3
# Create a scatter plot using hvplot.scatter to present the clustered data about cryptocurrencies
# having x="TotalCoinsMined" and y="TotalCoinSupply" 
# to contrast the number of available coins versus the total number of mined coins.

clustered_df.hvplot.scatter(x="TotalCoinsMined", y="TotalCoinSupply", by="class", hover_cols=["CoinName"])

:NdOverlay   [class]
   :Scatter   [TotalCoinsMined]   (TotalCoinSupply,CoinName)